File: test/integration/features/subscription/open_rescue_window_test.dart
// test/integration/features/subscription/open_rescue_window_test.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:fake_cloud_firestore/fake_cloud_firestore.dart';
import 'package:flutter_test/flutter_test.dart';

Future<void> simulateOpenRescueWindow(
  FakeFirebaseFirestore db,
  DateTime now,
) async {
  final threeDaysFromNow = now.add(const Duration(days: 3));

  final snapshot = await db.collection('homes')
      .where('premiumStatus', isEqualTo: 'cancelled_pending_end')
      .get();

  for (final doc in snapshot.docs) {
    final data = doc.data();
    final endsAt = (data['premiumEndsAt'] as Timestamp?)?.toDate();
    if (endsAt == null || endsAt.isAfter(threeDaysFromNow)) continue;
    if (data['rescueFlags']?['isInRescue'] == true) continue;

    final daysLeft = endsAt.difference(now).inDays.clamp(0, 3);
    await doc.reference.update({'premiumStatus': 'rescue'});
    await db.collection('homes').doc(doc.id).collection('views').doc('dashboard').set({
      'rescueFlags': {'isInRescue': true, 'daysLeft': daysLeft},
    }, SetOptions(merge: true));
  }
}

void main() {
  late FakeFirebaseFirestore db;
  const homeId = 'home1';

  setUp(() async {
    db = FakeFirebaseFirestore();
  });

  test('hogar a 2 días → premiumStatus cambia a rescue', () async {
    final now = DateTime(2026, 4, 6);
    final endsAt = now.add(const Duration(days: 2));

    await db.collection('homes').doc(homeId).set({
      'premiumStatus': 'cancelled_pending_end',
      'premiumEndsAt': Timestamp.fromDate(endsAt),
      'rescueFlags': {'isInRescue': false},
    });

    await simulateOpenRescueWindow(db, now);

    final homeSnap = await db.collection('homes').doc(homeId).get();
    expect(homeSnap.data()!['premiumStatus'], 'rescue');

    final dashSnap = await db.collection('homes').doc(homeId).collection('views').doc('dashboard').get();
    expect(dashSnap.data()!['rescueFlags']['isInRescue'], true);
    expect(dashSnap.data()!['rescueFlags']['daysLeft'], 2);
  });

  test('hogar a 5 días → NO cambia a rescue', () async {
    final now = DateTime(2026, 4, 6);
    final endsAt = now.add(const Duration(days: 5));

    await db.collection('homes').doc(homeId).set({
      'premiumStatus': 'cancelled_pending_end',
      'premiumEndsAt': Timestamp.fromDate(endsAt),
      'rescueFlags': {'isInRescue': false},
    });

    await simulateOpenRescueWindow(db, now);

    final homeSnap = await db.collection('homes').doc(homeId).get();
    expect(homeSnap.data()!['premiumStatus'], 'cancelled_pending_end');
  });

  test('hogar ya en rescue → no se toca', () async {
    final now = DateTime(2026, 4, 6);
    final endsAt = now.add(const Duration(days: 1));

    await db.collection('homes').doc(homeId).set({
      'premiumStatus': 'rescue',
      'premiumEndsAt': Timestamp.fromDate(endsAt),
      'rescueFlags': {'isInRescue': true, 'daysLeft': 1},
    });

    await simulateOpenRescueWindow(db, now);

    final homeSnap = await db.collection('homes').doc(homeId).get();
    expect(homeSnap.data()!['premiumStatus'], 'rescue');
  });
}

File: functions/test/integration/open_rescue_window.test.ts
// functions/test/integration/open_rescue_window.test.ts
//
// openRescueWindow es un scheduled job (onSchedule).
// Lo testeamos llamando directamente al handler exportado.
// Usamos firebase-functions-test para hacer wrap del scheduled handler.

import * as admin from 'firebase-admin';
import * as functionsTest from 'firebase-functions-test';
import { cleanAll, createUser, createHome, getDb } from './helpers/setup';
import { openRescueWindow } from '../../src/entitlement/open_rescue_window';

const testEnv = functionsTest({ projectId: process.env.GCLOUD_PROJECT });
const wrapped = testEnv.wrap(openRescueWindow) as (req: any) => Promise<any>;

const HOME_NEAR = 'home-rescue-near';    // premiumEndsAt dentro de 2 días
const HOME_FAR = 'home-rescue-far';     // premiumEndsAt dentro de 10 días
const HOME_ALREADY = 'home-rescue-done'; // ya está en rescue
const OWNER = 'owner-rescue';

function daysFromNow(days: number): admin.firestore.Timestamp {
  return admin.firestore.Timestamp.fromDate(new Date(Date.now() + days * 24 * 60 * 60 * 1000));
}

beforeAll(async () => {
  await cleanAll();
  await createUser(OWNER);

  // HOME_NEAR: premiumStatus=cancelled_pending_end, endsAt en 2 días
  await createHome(HOME_NEAR, OWNER, {
    premiumStatus: 'cancelled_pending_end',
    premiumEndsAt: daysFromNow(2),
  });

  // HOME_FAR: premiumStatus=cancelled_pending_end, endsAt en 10 días (fuera de ventana)
  await createHome(HOME_FAR, OWNER, {
    premiumStatus: 'cancelled_pending_end',
    premiumEndsAt: daysFromNow(10),
  });

  // HOME_ALREADY: ya está en rescue
  await createHome(HOME_ALREADY, OWNER, {
    premiumStatus: 'cancelled_pending_end',
    premiumEndsAt: daysFromNow(1),
    rescueFlags: { isInRescue: true },
  });
});

afterAll(() => testEnv.cleanup());

describe('openRescueWindow — scheduled job', () => {
  it('hogar dentro de ventana de 3 días → premiumStatus cambia a rescue', async () => {
    await wrapped({}); // scheduled jobs no necesitan auth

    const homeDoc = await getDb().collection('homes').doc(HOME_NEAR).get();
    expect(homeDoc.data()!['premiumStatus']).toBe('rescue');
  });

  it('hogar dentro de ventana → dashboard tiene rescueFlags.isInRescue = true', async () => {
    const dashDoc = await getDb().collection('homes').doc(HOME_NEAR).collection('views').doc('dashboard').get();
    expect(dashDoc.data()?.['rescueFlags']?.['isInRescue']).toBe(true);
    expect(dashDoc.data()?.['rescueFlags']?.['daysLeft']).toBeGreaterThanOrEqual(0);
  });

  it('hogar fuera de ventana de 3 días → NO cambia a rescue', async () => {
    const homeDoc = await getDb().collection('homes').doc(HOME_FAR).get();
    expect(homeDoc.data()!['premiumStatus']).toBe('cancelled_pending_end');
  });

  it('hogar ya en rescue → NO se vuelve a procesar', async () => {
    const homeDoc = await getDb().collection('homes').doc(HOME_ALREADY).get();
    // premiumStatus sigue siendo cancelled_pending_end (no se procesó porque isInRescue=true)
    expect(homeDoc.data()!['premiumStatus']).toBe('cancelled_pending_end');
  });
});
